src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs

```diff
@@ -25,57 +28,61 @@ public class ToBase64Transform : ICryptoTransform
 
         public int TransformBlock(byte[] inputBuffer, int inputOffset, int inputCount, byte[] outputBuffer, int outputOffset)
         {
-            ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
+            // inputCount != InputBlockSize is not allowed
+            ThrowHelper.ValidateTransformBlock(inputBuffer, inputOffset, inputCount, InputBlockSize);
+            if (outputBuffer == null) ThrowHelper.ThrowArgumentNull(ThrowHelper.ExceptionArgument.outputBuffer);
 
             // For now, only convert 3 bytes to 4
-            byte[] tempBytes = ConvertToBase64(inputBuffer, inputOffset, 3);
+            Span<byte> input = inputBuffer.AsSpan(inputOffset, InputBlockSize);
+            Span<byte> output = outputBuffer.AsSpan(outputOffset, OutputBlockSize);
+
+            OperationStatus status = Base64.EncodeToUtf8(input, output, out int consumed, out int written, isFinalBlock: false);
+
+            if (written != OutputBlockSize)
+                ThrowHelper.ThrowCryptographicException();
 
-            Buffer.BlockCopy(tempBytes, 0, outputBuffer, outputOffset, tempBytes.Length);
-            return tempBytes.Length;
+            Debug.Assert(status == OperationStatus.NeedMoreData);
+            Debug.Assert(consumed == InputBlockSize);
+
+            return written;
         }
 
         public byte[] TransformFinalBlock(byte[] inputBuffer, int inputOffset, int inputCount)
         {
-            ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
+            // inputCount <= InputBlockSize is allowed
+            ThrowHelper.ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
 
             // Convert.ToBase64CharArray already does padding, so all we have to check is that
             // the inputCount wasn't 0
             if (inputCount == 0)
             {
                 return Array.Empty<byte>();
             }
+            else if (inputCount > InputBlockSize)
+            {
+                ThrowHelper.ThrowArgumentInvalidValue(ThrowHelper.ExceptionArgument.inputCount);
```

Comment:
Was this already throwing in this case, just obliquely?

Reply:
It didn't throw directly. In the original code there was
https://github.com/dotnet/corefx/blob/94c5f1384c9b6477d164050de964cf6dcf99b2a9/src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs#L52-L55
So if `inputCount` is `> 3` it would crash at `Convert.ToBase64CharArray` with
```
System.ArgumentOutOfRangeException
  HResult=0x80131502
  Message=Either offset did not refer to a position in the string, or there is an insufficient length of destination character array.
```

But I see now it throws a `ArgumentException`. Change the exception-type to `ArgumentOutOfRangeException`?

Reply:
> Change the exception-type to `ArgumentOutOfRangeException`?

Ah, already covered by https://github.com/dotnet/corefx/pull/38062#discussion_r290839539